clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -77,19 +79,26 @@
     private boolean retry;
     private boolean reopened;
 
-    public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, long createdMs) {
+    public ProducerBatch(final TopicPartition tp, final MemoryRecordsBuilder recordsBuilder, final long createdMs) {
         this(tp, recordsBuilder, createdMs, false);
     }
 
-    public ProducerBatch(TopicPartition tp, MemoryRecordsBuilder recordsBuilder, long createdMs, boolean isSplitBatch) {
+    public ProducerBatch(final TopicPartition tp, final MemoryRecordsBuilder recordsBuilder,
+                         final long createdMs, final boolean isSplitBatch) {
+        this(tp, recordsBuilder, createdMs, isSplitBatch, new ProduceRequestResult(tp), null);
+    }
+
+    ProducerBatch(final TopicPartition tp, final MemoryRecordsBuilder recordsBuilder, final long createdMs,
+                  final boolean isSplitBatch, final ProduceRequestResult produceFuture, final List<ProducerBatch> childrenProducerBatch) {
         this.createdMs = createdMs;
         this.lastAttemptMs = createdMs;
         this.recordsBuilder = recordsBuilder;
         this.topicPartition = tp;
         this.lastAppendTime = createdMs;
-        this.produceFuture = new ProduceRequestResult(topicPartition);
+        this.produceFuture = produceFuture;
         this.retry = false;
         this.isSplitBatch = isSplitBatch;
+        this.childrenProducerBatch = childrenProducerBatch;
```

Comment:
We have multiple null check for `childrenProducerBatch` which is not necessary, instead we could just reject here if the given `childrenProducerBatch` is null to ensure it's non-null.

Reply:
(If I got your idea correctly)
We were doing that previously, but in that case, we will be always creating an `ArrayList` reserving memory that may not be used in case a batch does not get splitted.

Reply:
The other way we could go is to create `new ArrayList(0)` if we think that all these null checks are a pain. I do think we should avoid creating a new ArrayList with pre-allocated  elements with every batch though. Either way we should probably leave a comment next to https://github.com/apache/kafka/pull/7877/files#diff-e735b14a0fe9e843e528c2ce721602cbR72 with our reasoning or we risk it being changed inadvertently.